File: src/main/java/szewek/fl/util/JavaUtils.java
package szewek.fl.util;

import kotlin.jvm.internal.Intrinsics;
import net.minecraftforge.fml.common.discovery.ASMDataTable;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Objects;
import java.util.Set;
import java.util.function.BiConsumer;

public final class JavaUtils {
	public static final JavaUtils INSTANCE;

	@Nullable
	public final Class getClassSafely(@Nonnull String name) {
		Class var2;
		try {
			var2 = Class.forName(name);
		} catch (ClassNotFoundException var4) {
			var2 = null;
		}

		return var2;
	}

	@Nullable
	public final Method getMethodSafely(@Nonnull Class c, @Nonnull String name, @Nonnull Class... cargs) {
		Method var4;
		try {
			var4 = c.getDeclaredMethod(name, (Class[]) Arrays.copyOf(cargs, cargs.length));
		} catch (NoSuchMethodException var6) {
			var4 = null;
		}

		return var4;
	}

	public final Object[] makeFilledArray(@Nonnull Object[] t, Object fill) {
		int l = t.length - 1;
		int i = 0;
		int var5 = l;
		if (i <= l) {
			while (true) {
				t[i] = fill;
				if (i == var5) {
					break;
				}

				++i;
			}
		}

		return t;
	}

	public final void eachAnnotatedClasses(@Nonnull ASMDataTable asdt, @Nonnull Class<Annotation> ac, @Nonnull BiConsumer<Annotation, Class> eachfn) {
		Set<ASMDataTable.ASMData> aset = asdt.getAll(ac.getName());

		for (ASMDataTable.ASMData data : aset) {
			String cn = data.getClassName();
			if (Objects.equals(cn, data.getObjectName())) {
				Intrinsics.checkExpressionValueIsNotNull(cn, "cn");
				Class c = this.getClassSafely(cn);
				if (c != null) {
					Annotation a = c.getAnnotation(ac);
					eachfn.accept(a, c);
				}
			}
		}

	}

	private JavaUtils() {}

	static {
		INSTANCE = new JavaUtils();
	}
}

File: src/main/java/szewek/fl/util/JavaUtils.kt
package szewek.fl.util

import net.minecraftforge.fml.common.discovery.ASMDataTable
import java.lang.reflect.Method
import java.util.function.BiConsumer

object JavaUtils {

    fun getClassSafely(name: String): Class<*>? = try {
        Class.forName(name)
    } catch (ignored: ClassNotFoundException) {
        null
    }

    fun getMethodSafely(c: Class<*>, name: String, vararg cargs: Class<*>): Method? = try {
        c.getDeclaredMethod(name, *cargs)
    } catch (ignored: NoSuchMethodException) {
        null
    }

    fun <T> makeFilledArray(t: Array<T>, fill: T): Array<T> {
        val l = t.size - 1
        for (i in 0..l) {
            t[i] = fill
        }
        return t
    }

    fun <A : Annotation> eachAnnotatedClasses(asdt: ASMDataTable, ac: Class<A>, eachfn: BiConsumer<A, Class<*>>) {
        val aset = asdt.getAll(ac.name)
        for (data in aset) {
            val cn = data.className
            if (cn != data.objectName) continue
            val c = getClassSafely(cn) ?: continue
            val a = c.getAnnotation(ac)
            eachfn.accept(a, c)
        }
    }
}
